### PR TITLE
Pygameviewer

### PR DESCRIPTION
**Pull request by ptersilie**
This PR adds a new treeviewer to Eco, based on rpythons dotviewer.

Things that need to be checked when reviewing this PR are:
- [x] Toggling whitespace on/off (Key: W)
- [x] Moving through the tree history (Key: Meta left/right)
- [x] Closing the treeviewer and reopening it without crashing it

### CODE REVIEW

**Comment by snim2** [comment]
BTW #141 is right in the middle of a really tricky set of merge conflicts, to bring that branch up to date with the current master. If you could wait until PR #141 has been merged before merging this PR, that would be really helpful.

**Comment by ptersilie** [comment]
@snim2 sure can do. How long until you are merging #141? As I need the treeviewer PR before I can continue on my other branches.

**Comment by snim2** [comment]
@ptersilie much appreciated. Work on the branch is finished and squashed, it just needs bringing up to date before I can merge it. I'm expecting to be able to do that today (but the merge conflicts are tricky).

File: lib/eco/incparser/astree.py [hunk 1]
```diff
@@ -262,6 +262,18 @@ def load(self, version):
...
+        if version is None:
+            return self.__getattribute__(attr)
+        version = int(version)
+        while version >= 0:
+            try:
+                return self.log[(attr, version)]
+            except KeyError:
+                version -= 1
+        raise Exception("Warning: reached zero", self, self.log)
+        return None
```
Comment (cfbolz) [inline lib/eco/incparser/astree.py]: the `return None` is dead code. Shouldn't this be a somewhat less general exception type?

Comment (ptersilie) [inline lib/eco/incparser/astree.py]: Ah, yes you are right. I'm actually not even sure what I want to do here, as in if I even want an exception here. I guess I go with a exception for now. Do you have a suggestion for a better exception?

File: lib/eco/pgviewer.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        if lookup == "<ws>":
+            label = "<ws> x%s" % len(label)
+        label = label.replace("\"", "\\\"")
+        label = label.replace("\\\\\"", "\\\\\\\"")
+        label = label.replace("\r", "\\n")
+        dotnode = pydot.Node("\"%s\"" % id(node), label='"%s"' % label, shape="box")
+
+        try:
+            changed = node.get_attr("changed", self.version)
+        except Exception:
```
Comment (cfbolz) [inline lib/eco/pgviewer.py]: catching an ignoring every kind of exception here hides errors, just catch a precise type.

File: lib/eco/pgviewer.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            c_node = self.add_node(node.symbol.parser, graph)
+            graph.add_edge(pydot.Edge(dotnode, c_node))
+
+        return dotnode
+
+def splitline(line, re_word = re.compile(r'[^\s"]\S*|["]["]|["].*?[^\\]["]')):
+    result = []
+    for word in re_word.findall(line):
+        if word.startswith('"'):
+            word = eval(word)
```
Comment (cfbolz) [inline lib/eco/pgviewer.py]: please don't use eval on user input ;-)

**Comment by cfbolz** [comment]
All the three things you wanted me to test work.

**Comment by cfbolz** [comment]
Maybe you can add the eco-specific key bindings to the in-viewer help (press h or ? to get it).

File: lib/eco/pgviewer.py (hunk 3 above)
Comment (ptersilie) [inline lib/eco/pgviewer.py]: This is not my code. ;)

File: lib/eco/pgviewer.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        except Exception:
+            label = node.symbol.name
+            lookup = node.lookup
+        if label == "\r":
+            label = "<return>"
+        if lookup == "<ws>":
+            label = "<ws> x%s" % len(label)
+        label = label.replace("\"", "\\\"")
+        label = label.replace("\\\\\"", "\\\\\\\"")
+        label = label.replace("\r", "\\n")
```
Comment (cfbolz) [inline lib/eco/pgviewer.py]: this looks strange, do you still need it?

**Comment by snim2** [comment]
PR #141 has been merged now, thanks for waiting.

**Comment by ptersilie** [comment]
I already added the keybindings to the help.

**Comment by ptersilie** [comment]
Merged master into branch

**Comment by ltratt** [comment]
I think we can (and should) fix the `eval` issue by doing something like:

``` python
if word.startswith("\""):
  assert word[-1] == "\""
  word = word[1:-1]
```

**Comment by ptersilie** [comment]
If that is all this `eval` does, then yes.

**Comment by ltratt** [comment]
Please check, but I _think_ that's purely what it's aiming to do. [Of course, it can also eval any code, so you could pass `"blah" + os.system('rm -r /')` and it would happily wipe your hard disk!]

**Comment by ptersilie** [comment]
Seems to be working and I agree with your fix.

**Comment by ltratt** [comment]
OK, excellent!

One thing I'm unconvinced by is including a font: DroidSans is widely available (even I have it installed!) and we should just tell people to install it. If you feel really strongly about using that particular font, you can even print a warning saying "DroidSans not found: install it from X", but I don't think we want to start shipping ttf files.

**Comment by ptersilie** [comment]
I didn't write this code. I don't feel strongly about anything. ;) That being said I agree that not shipping a font is probably a good idea.

**Comment by ltratt** [comment]
Hehe :) OK, let's just delete it, and not worry about a warning then.

**Comment by ptersilie** [comment]
Removed the fonts and added checks to see if droid is installed and if not fall back to the default system font

**Comment by ltratt** [comment]
If I press "r" for "reload" I get:

```
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/home/ltratt/src/eco/lib/eco/pgviewer.py", line 15, in run
    self.gl.display()
  File "/home/ltratt/src/eco/lib/eco/dotviewer/drawgraph.py", line 81, in display
    self.get_display().run()
  File "/home/ltratt/src/eco/lib/eco/dotviewer/graphdisplay.py", line 749, in run
    self.run1()
  File "/home/ltratt/src/eco/lib/eco/dotviewer/graphdisplay.py", line 743, in run1
    self.process_event(EventQueue.pop(0))
  File "/home/ltratt/src/eco/lib/eco/dotviewer/graphdisplay.py", line 631, in process_event
    method(event)
  File "/home/ltratt/src/eco/lib/eco/dotviewer/graphdisplay.py", line 690, in process_KeyDown
    method(*args)
  File "/home/ltratt/src/eco/lib/eco/dotviewer/graphdisplay.py", line 431, in reload
    self.layout.request_reload()
AttributeError: GraphLayout instance has no attribute 'request_reload'
```

**Comment by ltratt** [comment]
If I load in `rng.eco`, I get one language box's root node in green, but no other node is coloured the same way. Is this significant?

**Comment by ptersilie** [comment]
The green thing means the node has changes. I never clear up the changed flag of the root (which I probably should out of consistency) but it never matter as the root is always changed anyways.

**Comment by ltratt** [comment]
In this case, the root root (if that makes sense) isn't green, but some random language box's root is green.

**Comment by ptersilie** [comment]
Yes, it doesn't matter at the moment that some root nodes are changed. For consistency reasons they shouldn't be, but since we always go down that path it doesn't change anything either. This is also something in the parser that has nothing to do with this PR.

**Comment by ptersilie** [comment]
I fixed the traceback on reload issue tho

**Comment by ltratt** [comment]
OK, fair enough. So once the tests have passed, let's squash, then I think we can merge.

**Pull request merged**